src/mtcp/mtcp_restart.c

```diff
@@ -632,19 +683,68 @@ main(int argc, char *argv[], char **environ)
+     *    lower than vDSO. However, on CentOS VM, we do not see any vvar section
+     *    at all on restart.
+     * Fix: There should not be any assumptions. We should be checking procmaps
+     *      to get the location of both vvar and vdso regions. And, move only if
+     *      they exist in procmaps. This code can be made generic.
+     * Problem 2: These mapped regions will be removed in
+     *    unmap_memory_areas_and_restore_vdso function as mtcp unmap all its
+     *    regions. And, we'll not be able to move it back to original program's
+     *    vdso region.
+     * What-we-do:
+     * 1. move mtcp's vdso and vvar regions temporarily at a temporary location
+     * 2. unmap all mtcp's memory regions
+     * 3. move from temp locations to ckpt'ed process' vdso vvar location.
+     * Fix: We should be bookeeping regions that we don't want MTCP to unmap 
+     */
+#ifdef WORKING:
+    setupTempRegions();
+#endif
+    Area area;
+    int mapsfd = mtcp_sys_open2("/proc/self/maps", O_RDONLY);
+    if (mapsfd < 0) {
+      MTCP_PRINTF("error opening /proc/self/maps; errno: %d\n", mtcp_sys_errno);
+      mtcp_abort();
+    }
+
+    while (mtcp_readmapsline(mapsfd, &area)) {
+      void *rc = 0x0;
+      if (mtcp_strcmp(area.name, "[vvar]") == 0) {
+        rc = mtcp_sys_mremap(area.addr, area.size, area.size,
+	                     MREMAP_FIXED | MREMAP_MAYMOVE, vvarStartTmp);
+      }
+      if (mtcp_strcmp(area.name, "[vdso]") == 0) {
+        rc = mtcp_sys_mremap(area.addr, area.size, area.size,
+                             MREMAP_FIXED | MREMAP_MAYMOVE, vdsoStartTmp);
+	mtcp_printf("%p\n", vdsoStartTmp);
+      }
+      if (rc == MAP_FAILED) {
+        MTCP_PRINTF("mtcp_restart failed: "
+                    "gdb --mpi \"DMTCP_ROOT/bin/mtcp_restart\" to debug.\n");
+      }
+    }
+    mtcp_sys_close(mapsfd);
+
+#if 0
     char *vvarStart = vdsoStart - 3 * PAGESIZE;
-    void *rc1 = mtcp_sys_mremap(vvarStart, 3*PAGESIZE, 3*PAGESIZE,
-		                MREMAP_FIXED | MREMAP_MAYMOVE, vvarStartTmp);
+//    void *rc1 = mtcp_sys_mremap(vvarStart, 3*PAGESIZE, 3*PAGESIZE,
+//		                MREMAP_FIXED | MREMAP_MAYMOVE, vvarStartTmp);
     void *rc2 = mtcp_sys_mremap(vdsoStart, 2*PAGESIZE, 2*PAGESIZE,
 		                MREMAP_FIXED | MREMAP_MAYMOVE, vdsoStartTmp);
     if (rc2 == MAP_FAILED) { // vdso segment can be one page or two pages.
       rc2 = mtcp_sys_mremap(vdsoStart, 1*PAGESIZE, 1*PAGESIZE,
 		            MREMAP_FIXED | MREMAP_MAYMOVE, vdsoStartTmp);
     }
+    void * rc1 = 0x0; // @Illio: this should go away in the final code
     if (rc1 == MAP_FAILED || rc2 == MAP_FAILED) {
       MTCP_PRINTF("mtcp_restart failed: "
 		  "gdb --mpi \"DMTCP_ROOT/bin/mtcp_restart\" to debug.\n");
     }
+#endif
+
     // Now that we moved vdso/vvar, we need to update the vdso address
     //   in the auxv vector.  This will be needed when the lower half
     //   starts up:  initializeLowerHalf() will be called from splitProcess().
```

Comment:
this (and line 751, i can't include it in the highlight) should also be obsolete. `mygetauxval` and `mysetauxval` are only called here.